x/protorev/protorev.md

```diff
@@ -0,0 +1,697 @@
+
+```go
+// TokenPairArbRoutes tracks all of the hot routes for a given pair of tokens
+message TokenPairArbRoutes {
+  option (gogoproto.equal) = true;
+
+  // Stores all of the possible hot paths for a given pair of tokens
+  repeated Route arb_routes = 1;
+  // Token denomination of the first asset
+  string token_in = 2;
+  // Token denomination of the second asset
+  string token_out = 3;
+}
+
+// Route is a hot route for a given pair of tokens
+message Route {
+  option (gogoproto.equal) = true;
+
+  // The pool IDs that are travered in the directed cyclic graph (traversed left
+  // -> right)
+  repeated Trade trades = 1;
+  // The step size that will be used to find the optimal swap amount in the
+  // binary search
+  string step_size = 2 [
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.nullable) = true
+  ];
+}
+
+// Trade is a single trade in a route
+message Trade {
+  option (gogoproto.equal) = true;
+
+  // The pool IDs that are travered in the directed cyclic graph (traversed left
+  // -> right)
+  uint64 pool = 1;
+  // The denom of token A that is traded
+  string token_in = 2;
+  // The denom of token B that is traded
+  string token_out = 3;
+}
+```
+
+### DenomPairToPool
+
+DenomPairToPool takes in a base denomination (read below) – denom that is used to build routes (ex. osmo, atom, usdc) – and a denom to match (akash, juno) and returns the highest liquidity pool id between the pair of denominations. For example, an input might look like (osmo, juno) —> poolID: 5. This store is directly tied to the highest liquidity method (described in state transitions below). Each base denomination is going to have its own set of denominations it maps to.
+
+### BaseDenoms
+
+BaseDenoms are the denominations that are used to build the highest liquidity routes. This will be configurable by the admin account, but will always maintain at least `uosmo` as a base denom. A base denom just means the denomination that will be used to start and end a cyclic arbitrage route. Base denoms can be added on as needed basis. 
+
+***NOTE***: BaseDenoms do have a priority that is directly tied down to the order in the list of base denoms that are used i.e. BaseDenoms that are closer to the front of the list will likely be simulated and executed more often than those later in the list. This is done by design so that we can prioritize certain denoms over others in order to simulate and execute the most profitable trades.
```

Comment:
Priority for a denomination is vaguely tied down to how frequently we might see an arbitrage opportunity for that denomination. In practice, we will be conducting more research on the metrics we might want to use to determine which denominations should have priority over others.

Two of the more basic metrics might be something like hit rate - how often a denomination creates an arbitrage opportunity for every time we see it in a swap - and fill rate - how profitable each swap was by denomination.

More generally though this is an open ended problem. We wanted a way to be able to prioritize certain denominations over others in the case where we can only simulate a handful of routes. For example, lets say `osmo` ends up being the most profitable arbitrage denomination with `atom` being a distant second. In the case where we can only simulate and execute one route close to the end of the block because we have already exhausted all of the compute - pool points - in previous transcations, we would choose to simulate `osmo` because it might have a higher probability of generating a profitable route.